lib/sentry/client_error.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Sentry.ClientError do
+  @moduledoc """
+  An exception struct that represents an error returned by Sentry.
+
+  This struct is designed to manage and handle errors originating from operations
+  in the client application. The `:reason` field signifies the cause of the error
+  as an atom or tuple.
+
+  To raise instances of this exception, you can use `Kernel.raise/1`. When crafting
+  formatted error messages for purposes such as logging or presentation, consider
+  leveraging `Exception.message/1`.
+  """
+
+  @moduledoc since: "10.7.0"
+
+  @doc """
+  The exception struct for a Sentry error.
+  """
+  defexception [:reason]
+
+  @typedoc """
+  The type for a Sentry error exception.
+  """
+  @type t() :: %__MODULE__{reason: term}
+
+  @doc false
+  @spec new(term) :: t
+  def new(reason) do
+    %__MODULE__{reason: reason}
+  end
+
+  @impl true
+  def message(%__MODULE__{reason: reason}) do
+    "Request failure reason: #{format(reason)}"
+  end
+
+  defp format(:too_many_retries) do
+    "Sentry responded with status 429 - Too Many Requests"
+  end
+
+  defp format({:invalid_json, reason}) do
+    "Invalid JSON -> #{inspect(reason)}"
```

Comment:
This should be a `Jason.DecodeError`, right? So we can do

```suggestion
    "Invalid JSON: #{Exception.message(reason)}"
```

Reply:
hmmm actually I think I'm wrong here. This exception get returned on the :request_failure tuple, not with :invalid_json. After following the path it looks like the error produced for this atom is one of these:

`  def __validate_json_library__(nil) do
    {:error, "nil is not a valid value for the :json_library option"}
  end

  def __validate_json_library__(mod) when is_atom(mod) do
    try do
      with {:ok, %{}} <- mod.decode("{}"),
           {:ok, "{}"} <- mod.encode(%{}) do
        {:ok, mod}
      else
        _ ->
          {:error,
           "configured :json_library #{inspect(mod)} does not implement decode/1 and encode/1"}
      end
    rescue
      UndefinedFunctionError ->
        {:error,
         """
         configured :json_library #{inspect(mod)} is not available or does not implement decode/1 and encode/1.
         Do you need to add #{inspect(mod)} to your mix.exs?
         """}
    end
  end

  def __validate_json_library__(other) do
    {:error, "expected :json_library to be a module, got: #{inspect(other)}"}
  end`

Is that correct?

Reply:
I don't think that's it, that's something we do when the `:sentry` application starts to validate that the JSON library chosen by the user quacks like a JSON library indeed.

IIRC we return `:invalid_json` when the response from the Sentry server is not valid JSON, don't we?

https://github.com/getsentry/sentry-elixir/blob/fba608c72a403cb0b21e9ebd8cd2cd313a1e3ecb/lib/sentry/transport.ex#L23-L30